### ISSUE TITLE
wmake crashes with a buffer overflow

### ISSUE THREAD

**Issue by Jookia** [joncampbell123/dosbox-x#3388]
### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

### Have you checked that no other similar bug report(s) already exists?

- [X] I have searched and didn't find any similar issues.

### What operating system(s) this bug have occurred on?

Ubuntu 21.10 amd64, qemu

### What version(s) of DOSBox-X have this bug?

master

### Describe the bug

After #3303 was merged, wmake sometimes crashes now.

### Expected behavior

Wmake doesn't crash.

### Steps to reproduce the behaviour

Try running wmake or compiling something.

### Additional context

Backtrace:

```** buffer overflow detected ***: terminated

Thread 1 "dosbox-x" received signal SIGABRT, Aborted.
__pthread_kill_implementation (no_tid=0, signo=6, threadid=140263273332544) at pthread_kill.c:44
44	pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140263273332544) at pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140263273332544) at pthread_kill.c:80
#2  __GI___pthread_kill (threadid=140263273332544, signo=signo@entry=6) at pthread_kill.c:91
#3  0x00007f9197476476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f919745c7b7 in __GI_abort () at abort.c:79
#5  0x00007f91974bd606 in __libc_message (action=action@entry=do_abort, 
    fmt=fmt@entry=0x7f919760eef4 "*** %s ***: terminated\n") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007f919756935a in __GI___fortify_fail (msg=msg@entry=0x7f919760ee9a "buffer overflow detected")
    at fortify_fail.c:26
#7  0x00007f9197567c76 in __GI___chk_fail () at chk_fail.c:28
#8  0x00007f91975676a9 in __strncpy_chk (s1=s1@entry=0x5582cee5dba0 <appargs> "", 
    s2=s2@entry=0x7ffea04f9721 "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016C:\\WATCOM\\BINW\\dos4gw.exe", n=<optimised out>, s1len=s1len@entry=127) at strncpy_chk.c:26
#9  0x00005582cdf5ae40 in strncpy (__len=<optimised out>, 
    __src=0x7ffea04f9721 "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016C:\\WATCOM\\BINW\\dos4gw.exe", __dest=0x5582cee5dba0 <appargs> "")
    at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:95
#10 DOS_21Handler () at dos.cpp:2212
#11 0x00005582cdeb97b4 in Normal_Loop () at dosbox.cpp:371
#12 0x00005582cdeb99b6 in DOSBOX_RunMachine () at dosbox.cpp:630
#13 0x00005582ce317ae3 in CALLBACK_RunRealInt (intnum=intnum@entry=33 '!') at callback.cpp:209
#14 0x00005582ce07bd0f in DOS_Shell::Execute (this=this@entry=0x5582d49ecb80, 
    name=name@entry=0x7ffea04fb410 "wmake", args=args@entry=0x7ffea04fc785 "") at shell_misc.cpp:1495
#15 0x00005582ce066bba in DOS_Shell::DoCommand (this=this@entry=0x5582d49ecb80, line=0x7ffea04fc785 "", 
    line@entry=0x7ffea04fc780 "wmake") at shell_cmds.cpp:276
#16 0x00005582ce052ea1 in DOS_Shell::ParseLine (this=0x5582d49ecb80, line=0x7ffea04fc780 "wmake")
    at shell.cpp:483
#17 0x00005582ce05a813 in DOS_Shell::Run (this=0x5582d49ecb80) at shell.cpp:959
#18 0x00005582ce059ba6 in SHELL_Run () at shell.cpp:1828
#19 0x00005582ce10decc in VM_Boot_DOSBox_Kernel () at sdlmain.cpp:7125
#20 0x00005582ce0d02d2 in BIOS::cb_bios_boot__func () at bios.cpp:9587
#21 0x00005582cdeb97b4 in Normal_Loop () at dosbox.cpp:371
#22 0x00005582cdeb99b6 in DOSBOX_RunMachine () at dosbox.cpp:630
--Type <RET> for more, q to quit, c to continue without paging--q
Quit
(gdb) frame 10
#10 DOS_21Handler () at dos.cpp:2212
in2212	                    strncpy(appargs, ctail.buffer, strlen(ctail.buffer) <= sizeof(appargs) ? strlen(ctail.buffer) : sizeof(appargs));
(gdb) info locals
block = {<MemStruct> = {pt = 70944}, exec = {envseg = 4434, cmdtail = 2425360722, fcb1 = 144, fcb2 = 0, 
    initsssp = 1364393984, initcsip = 3092731474}, overlay = {loadseg = 4434, relocation = 4434}}
ctail = {count = 84 'T', 
  buffer = "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016"}
unmask_irq0 = false
name1 = "C:\\WATCOM\\BINW\\dos4gw.exe\000/ՂU\000\000\320\002\000\000\000\000\000\000\000\234f\274\261\250\336c\260\263\214ՂU\000\000@\372\230ԂU\000\000\000\000\000\000\000\000\000\000\001", '\000' <repeats 15 times>, "X\372\230ԂU\000\000\030\230O\240\376\177\000\000\237\016.\227\221\177\000\000\034\000\000\000\000\000\000\000\374\v\243\227\221\177\000\000P\001\232ԂU\000\000\000\234f\274\261\250\336c\021\000@\003\000\000\000\000\bO\242\227\221\177\000\000\340\230O\240\376\177\000\000\034\000\000\000\000\000\000\000\320\347\230ԂU\000\000P\001\232ԂU\000\000\340\230O\240\376\177\000\000\336uʗ\221\177\000\000"...
name2 = " name bot.exe file obj\\tcplib\\*.obj,obj\\bot\\*.obj\rbj\000\\*.obj,obj\\bot\\*.obj\rbj\000U\000\000\021\000@\003\000\000\000\000/\001\000\000\000\000\000\000\236\334X\000\000\000\000\000\000\000\000\000\241\001", '\000' <repeats 26 times>, "\320n?ՂU\000\000\300\334d\227\221\177", '\000' <repeats 18 times>, "\236\266\022\215\207U\000\000mount folder read-only\000\000\f"...
time_start = 0
(gdb) bt
#0  __pthread_kill_implementation (no_tid=0, signo=6, threadid=140263273332544) at pthread_kill.c:44
#1  __pthread_kill_internal (signo=6, threadid=140263273332544) at pthread_kill.c:80
#2  __GI___pthread_kill (threadid=140263273332544, signo=signo@entry=6) at pthread_kill.c:91
#3  0x00007f9197476476 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
#4  0x00007f919745c7b7 in __GI_abort () at abort.c:79
#5  0x00007f91974bd606 in __libc_message (action=action@entry=do_abort, 
    fmt=fmt@entry=0x7f919760eef4 "*** %s ***: terminated\n") at ../sysdeps/posix/libc_fatal.c:155
#6  0x00007f919756935a in __GI___fortify_fail (msg=msg@entry=0x7f919760ee9a "buffer overflow detected")
    at fortify_fail.c:26
#7  0x00007f9197567c76 in __GI___chk_fail () at chk_fail.c:28
#8  0x00007f91975676a9 in __strncpy_chk (s1=s1@entry=0x5582cee5dba0 <appargs> "", 
    s2=s2@entry=0x7ffea04f9721 "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016C:\\WATCOM\\BINW\\dos4gw.exe", n=<optimised out>, s1len=s1len@entry=127) at strncpy_chk.c:26
#9  0x00005582cdf5ae40 in strncpy (__len=<optimised out>, 
    __src=0x7ffea04f9721 "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016C:\\WATCOM\\BINW\\dos4gw.exe", __dest=0x5582cee5dba0 <appargs> "")
    at /usr/include/x86_64-linux-gnu/bits/string_fortified.h:95
#10 DOS_21Handler () at dos.cpp:2212
#11 0x00005582cdeb97b4 in Normal_Loop () at dosbox.cpp:371
#12 0x00005582cdeb99b6 in DOSBOX_RunMachine () at dosbox.cpp:630
#13 0x00005582ce317ae3 in CALLBACK_RunRealInt (intnum=intnum@entry=33 '!') at callback.cpp:209
#14 0x00005582ce07bd0f in DOS_Shell::Execute (this=this@entry=0x5582d49ecb80, 
    name=name@entry=0x7ffea04fb410 "wmake", args=args@entry=0x7ffea04fc785 "") at shell_misc.cpp:1495
#15 0x00005582ce066bba in DOS_Shell::DoCommand (this=this@entry=0x5582d49ecb80, line=0x7ffea04fc785 "", 
    line@entry=0x7ffea04fc780 "wmake") at shell_cmds.cpp:276
#16 0x00005582ce052ea1 in DOS_Shell::ParseLine (this=0x5582d49ecb80, line=0x7ffea04fc780 "wmake")
    at shell.cpp:483
#17 0x00005582ce05a813 in DOS_Shell::Run (this=0x5582d49ecb80) at shell.cpp:959
#18 0x00005582ce059ba6 in SHELL_Run () at shell.cpp:1828
#19 0x00005582ce10decc in VM_Boot_DOSBox_Kernel () at sdlmain.cpp:7125
#20 0x00005582ce0d02d2 in BIOS::cb_bios_boot__func () at bios.cpp:9587
#21 0x00005582cdeb97b4 in Normal_Loop () at dosbox.cpp:371
#22 0x00005582cdeb99b6 in DOSBOX_RunMachine () at dosbox.cpp:630
--Type <RET> for more, q to quit, c to continue without paging--
#23 0x00005582cde6e16b in main (argc=<optimised out>, argv=0x7ffea04fe928) at sdlmain.cpp:8925
(gdb) frame 10
#10 DOS_21Handler () at dos.cpp:2212
2212	                    strncpy(appargs, ctail.buffer, strlen(ctail.buffer) <= sizeof(appargs) ? strlen(ctail.buffer) : sizeof(appargs));
(gdb) info locals
block = {<MemStruct> = {pt = 70944}, exec = {envseg = 4434, cmdtail = 2425360722, fcb1 = 144, fcb2 = 0, 
    initsssp = 1364393984, initcsip = 3092731474}, overlay = {loadseg = 4434, relocation = 4434}}
ctail = {count = 84 'T', 
  buffer = "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016"}
unmask_irq0 = false
name1 = "C:\\WATCOM\\BINW\\dos4gw.exe\000/ՂU\000\000\320\002\000\000\000\000\000\000\000\234f\274\261\250\336c\260\263\214ՂU\000\000@\372\230ԂU\000\000\000\000\000\000\000\000\000\000\001", '\000' <repeats 15 times>, "X\372\230ԂU\000\000\030\230O\240\376\177\000\000\237\016.\227\221\177\000\000\034\000\000\000\000\000\000\000\374\v\243\227\221\177\000\000P\001\232ԂU\000\000\000\234f\274\261\250\336c\021\000@\003\000\000\000\000\bO\242\227\221\177\000\000\340\230O\240\376\177\000\000\034\000\000\000\000\000\000\000\320\347\230ԂU\000\000P\001\232ԂU\000\000\340\230O\240\376\177\000\000\336uʗ\221\177\000\000"...
name2 = " name bot.exe file obj\\tcplib\\*.obj,obj\\bot\\*.obj\rbj\000\\*.obj,obj\\bot\\*.obj\rbj\000U\000\000\021\000@\003\000\000\000\000/\001\000\000\000\000\000\000\236\334X\000\000\000\000\000\000\000\000\000\241\001", '\000' <repeats 26 times>, "\320n?ՂU\000\000\300\334d\227\221\177", '\000' <repeats 18 times>, "\236\266\022\215\207U\000\000mount folder read-only\000\000\f"...
time_start = 0
(gdb) p appargs
$1 = '\000' <repeats 126 times>
(gdb) p ctail.buffer
$2 = "A\016XE\016\\E\016`E\016Pw\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vj\016TD\016XE\016\\E\016`L\016TC\016XB\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016\004A\vC\016\\H\016`H\016P\002M\016TA\016XG\016\\C\016`K\016PH\n\016\024A\303\016\020A\306\016\fA\307\016\bA\305\016"
```

**Comment by Wengier** [comment]
@maron2000 may want to look at this.

**Comment by maron2000** [comment]
I can't try wmake but may be worth trying the following patch to my previous PR.
```
//strncpy(appargs, ctail.buffer, strlen(ctail.buffer) <= sizeof(appargs) ? strlen(ctail.buffer) : sizeof(appargs)); //old
strncpy(appargs, ctail.buffer, strlen(ctail.buffer) <= sizeof(appargs) -1 ? strlen(ctail.buffer) : sizeof(appargs) -1); 
```

**Comment by Jookia** [comment]
I'll give that a shot and have a look at the code to see if there's a
nicer fix waiting to be found.

**Comment by Jookia** [comment]
Alright, I had a look at the code and learned a few things:

- We're copying the appname/appargs for HostAppRun so we can run Windows applications from DOS
- HostAppRun expects appname/appargs to be some kind of string
- appname/appargs is copied from DOS memory directly and may be buggy
- ctail.buffer and appargs are 127 bytes long
- ctail.count is a uint8_t, allowing users to go say the length is longer than the buffer size

The original code here:

```
                    strcpy(appname, name1);
                    strncpy(appargs, ctail.buffer, ctail.count); 
                    *(appargs+ctail.count)=0;
```

- strcpy should work because appname and name1 are both DOSNAMEBUF+2+DOS_NAMELENGTH_ASCII
- strncpy may copy past the buffers

The fix you implemented is to try and read ctail.buffer then clamp it to appargs. This is crashing because of the issue you found, and your code you just pasted there should fix it.

But this may still read past ctail.buffer if there's no null terminator in it, which is also a buffer overflow. It just might not error. ;)

So I sent a PR #3389 that will just copy the entire buffer and null terminate it ourselves.

**Comment by maron2000** [comment]
@Jookia
Thanks for your investigation. I have no reason to object your PR as long as it is OK running wmake.
I haven't tried but I assume your PR should be consistent to fix Issue #3285 as well.
